### PR TITLE
[Ready for Audit] Create ERC721 Lockable Module

### PR DESCRIPTION
**Pull request by darkty0x**

### CODE REVIEW

File: erc721-lockable/contracts/base/ERC721Lockable.sol [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+
+        _lock(address(0), to, tokenId, expired);
+    }
+
+    /**
+     * @dev See {ERC721-_burn}. This override additionally clears the lock approvals for the token.
+     */
+    function _burn(uint256 tokenId) internal virtual override {
+        address owner = ERC721.ownerOf(tokenId);
+        super._burn(tokenId);
```
Comment (deep-quality-dev) [inline erc721-lockable/contracts/base/ERC721Lockable.sol:304]: This will check if this `tokenId` was already locked before burning, correct?

Comment (kazunetakeda25) [reply erc721-lockable/contracts/base/ERC721Lockable.sol:304]: Yes, correct @deep-quality-dev

**Review by deep-quality-dev — COMMENTED** [review]
Overall, Looks good to me, do we check lock status before approval?
- isApprovedForAll
- setApprovalForAll
- _beforeApprove

**Comment by kazunetakeda25** [comment]
No, we don't. They will be just ERC721 methods and they work as usual.
But instead, we do hae isLockAPprovedForAll method, and they will check lock status before approval.

**Comment by deep-quality-dev** [comment]
> > Overall, Looks good to me, do we check lock status before approval?
> > 
> > * isApprovedForAll
> > * setApprovalForAll
> > * _beforeApprove
> 
> No, we don't. They will be just ERC721 methods and they work as usual. But instead, we do hae isLockAPprovedForAll method, and they will check lock status before approval.

That's fine, it's up to requirements, in some cases, it requires to check lock status while approval. Your current lockable contract is much flexible then. Great

**Review by deep-quality-dev — APPROVED** [review]

File: erc721-lockable/contracts/base/ERC721Lockable.sol [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+ * @notice ERC721 Token that can be locked for a certain period and cannot be transferred.
+ * This is designed for a non-escrow staking contract that comes later to lock a user's NFT
+ * while still letting them keep it in their wallet.
+ * This extension can ensure the security of user tokens during the staking period.
+ * If the nft lending protocol is compatible with this extension, the trouble caused by the NFT
+ * airdrop can be avoided, because the airdrop is still in the user's wallet
+ * @dev Implementation ERC721 Lockable Token
+ * @author kazunetakeda25
+ */
+abstract contract ERC721Lockable is ERC721, IERC721Lockable, IBunzz, Ownable {
```
Comment (solcmoon) [inline erc721-lockable/contracts/base/ERC721Lockable.sol:22]: If this is main contract, please remove `abstract`

Comment (solcmoon) [reply erc721-lockable/contracts/base/ERC721Lockable.sol:22]: and move this contract into root path

File: erc721-lockable/contracts/base/ERC721Lockable.sol [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+    mapping(address => mapping(address => bool)) private _lockOperatorApprovals;
+
+    /**
+     * @dev Connect to other contracts
+     */
+    function connectToOtherContracts(address[] calldata _contracts)
+        public
+        override
+        onlyOwner
+    {}
```
Comment (solcmoon) [inline erc721-lockable/contracts/base/ERC721Lockable.sol:38]: if you don't need this function, please remove this function and `IBunzz` interface

**Pull request closed without merging**